### DISCUSSION TITLE
Add page render hook to allow integrations to post-process and validate rendered HTML code

### DISCUSSION THREAD

**Discussion by hippotastic** [withastro/roadmap#169] [Proposal]
# Goals & use cases
As a developer, I want to be able to extend Astro's capabilities by **post-processing and validating the rendered HTML code of my pages**.

Here are some use cases that are currently supported by other SSGs/frameworks, but not by Astro (yet!):
- Prettify HTML output of `dev`
- Minify HTML output of `build`
- Process all links on the page to make them relative / absolute / point to a given domain
- Automatically add attributes like `rel="noopener noreferrer"` to all external links
- Ensure that all images on the page have their dimensions specified, a srcset defined etc.
- Validate the existence of SEO-relevant tags on all public-facing pages (e.g. enforcing all such pages to have a page `title` with a proper length, a valid `description`, the correct canonical link, JSON-LD WebPage / Article schemas etc.), and if these checks fail, print a console error in `dev` mode or fail the `build`
- ...and much more 😄

# Alternatives (what I've tried so far to do this in Astro)
- **Using the currently existing Integration API hooks provided by Astro:** Unfortunately, there is no Integration API hook that gets called per rendered page and that allows to transform the HTML code, and the only existing hook that has access to the rendered pages (`astro:build:done`) only gets called *after* all files have already been written to disk. Also, there is no such hook that gets called per rendered page during `dev` mode.
- **Writing a Vite plugin:** Unfortunately, this approach also doesn't work, because Vite plugins do not get called by Astro per page, but per bundle / chunk. The `transform` hook supported by Vite/rollup therefore also cannot be used, because it does not have access to the rendered page HTML, but only to JS bundles. It would be a lot of work to parse and modify the bundle code just to get access to the rendered pages. This is also the reason why many existing vite plugins don't really work well in Astro.
- **Chaining other CLI tools after `build`:** This is the currently "recommended" approach for minifying HTML output after `build`. I only consider this a workaround as it requires to perform additional globbing of the build output, and it's completely impossible in `dev` mode. Also, I assume this approach would fail in SSR mode as well.

# My solution idea / proposed change

## Initial solution idea
- Add a new hook function `runHookRenderPage` with an accompanying `astro:render:page` config setting to [astro/src/integrations/index.ts](https://github.com/withastro/astro/blob/main/packages/astro/src/integrations/index.ts).
- For `build` mode, call the new hook function from the `generatePath` function [before this line in astro/src/core/build/generate.ts](https://github.com/withastro/astro/blob/3c5715a74030f04a3b3185ee6b89f7184e1cef6b/packages/astro/src/core/build/generate.ts#L252).
- For `dev` mode, call the new hook function from the `handleRequest` function [in this block of astro/src/vite-plugin-astro-server](https://github.com/withastro/astro/blob/3c5715a74030f04a3b3185ee6b89f7184e1cef6b/packages/astro/src/vite-plugin-astro-server/index.ts#L279).
- Ensure that `runHookRenderPage` receives the Astro config, the path of the currently rendered page route, and the current body and status code.
- In `runHookRenderPage`, call all `astro:render:page` hook functions registered by Integrations one by one, allowing them to modify the rendered HTML code (and maybe also the status code, if that seems to be useful), and passing the modified result to the next Integration, until passing the final result back to the calling code, which can then either write it to a file (during `build`) or send it to the browser (during `dev`).

I've implemented the solution outlined above on my local machine, and it works well both during `dev` and `build`.

## Discussions about this solution idea so far
- I've had a very nice chat with @matthewp about this idea in the `#dev` Discord channel, who gave my use cases a thumbs up, and also proposed to change the solution idea to call the hook function from one central place in order to ensure it also works well for SSR. He proposed to call the hook from the `render` function [before this line in astro/src/core/render/core.ts](https://github.com/withastro/astro/blob/main/packages/astro/src/core/render/core.ts#L166).
- I like the idea of having one central place to call this hook from, but I have the problem that this function doesn't seem to have access to the Astro config, which makes it impossible for me to enumerate registered Integrations for their registered hooks and call any that registered `astro:render:page`.
- I don't know how to modify Astro to ensure that `render` always has proper access to the Astro config. If anyone could help on how to do that, I'd gladly submit a PR to implement this hook.

# What do you think?
First of all, thank you for reading through this long text! 💜

I'm open to any kind of feedback on this proposal and would love to contribute. Please just let me know if you support this feature idea, and if so, how we can proceed with it.

**Comment by hippotastic** [comment]
I have implemented the solution discussed with @matthewp. You can find my PR here: https://github.com/withastro/astro/pull/3050

(Update: PR is outdated, there's a new implementation based on Fred's input further below.)

**Comment by williamtetlow** [comment]
This looks great @hippotastic, I hope it gets approval!!

I think HTML prettify for dev vs minification for production builds should be a default and not something that needs to be configured but otherwise I'm super excited for this.

I want to use it to add [`critters`](https://github.com/GoogleChromeLabs/critters) integration 😄

**Comment by FredKSchott** [comment]
Thanks for kicking this off @hippotastic! This is 100% something that we want to add support for eventually, but it was kept out of the main implementation because it's more complex than the other hooks. A few things that we'll need to solve for:

1. The "render hook" will need to run in SSR. No other hook has this requirement. That means that it will need to be included in the final SSR build somehow. But how? Integrations themselves aren't designed to run in SSR, so how do you export just a single hook to run in SSR without also including the file that the hook lives in? This one is very tricky.
2. Since it needs to run in SSR, it can't have access to the config object, since the config object doesn't exist in SSR. This one is also very tricky.
3. We don't currently parse the HTML response into an AST for easy modifications. We don't actually need to solve for this immediately (an integration could just do it's own AST parse step if it needed it). But, we probably will eventually (otherwise each integration will end up doing its own AST parse, causing duplicate work).

I don't meant to be all doom and gloom though! This is definitely possible, and something I'm really excited to see!

It would help to think through these problems first. 1) is by far the toughest. Maybe we need something like `addAdapter()` and `addRenderer()`, which provide code that runs in SSR by giving it an entrypoint to import? Something like:

```
"astro:config:setup": ({addRenderHook}) => {
  addRenderHook({hookEntrypoint: '@astrojs/my-integration-name/hooks/render.js'})
})
// Question: how would you pass args to `hookEntrypoint`?
``

**Reply by williamtetlow** [reply]
For `3`, I wonder if that would be two different hooks 🤔 

You can access AST before it is printed in one hook or the HTML string after the AST has been printed in another

**Reply by matthewp** [reply]
Bundling astro.config.mjs is also possible and something I explored doing already. There could be other reasons why we might want to do that in the future. The downside is bundling a bunch of dev-only stuff as well.

**Reply by hippotastic** [reply]
When developing my first proof of concept for this discussion, I would have definitely preferred having access to the Astro config file from the SSR bundle. I did not know much about how SSR works in Astro and was surprised to learn that the bundling phase introduces a separation and cuts away access to the config.

In the end, I was able to add everything that I needed to the bundle with some extra work, but it was still an additional hurdle that forced me to think separately about the SSR use case. In in ideal world, SSR page generation should "just work" in the same way as SSG page generation, without requiring developers to cater specifically to the SSR scenario.

**Comment by hippotastic** [comment]
Thanks a lot for looking into this @FredKSchott! I underestimated the additional complexity of making this work for SSR. After your helpful explanations, I experimented with the Netlify SSR adapter and can see the challenges as well. Based on the contexts this is being called from, hooking into page rendering may indeed require a different kind of hook, and I like your `addRenderHook` idea a lot.

Regarding your question about passing args to `hookEntrypoint`, we could add an optional `options` parameter to the type that also defines `hookEntrypoint`. The only limitation I can see is that it must be serializable to JSON so that it can be stored in the generated SSR bundle.

**Comment by hippotastic** [comment]
See fully working implementation of the `addRenderHook` approach here:
https://github.com/hippotastic/astro/commit/58d77c1e126ae56fb7ec75c9102448c903628f41

I've also made it possible to pass options to the hook. Usage example:

## How to use a render hook integration as an end-user

You can use the integration like any other integration. Nothing new to learn for regular users of Astro.

**astro.config.mjs:**
```js
import { defineConfig } from 'astro/config'
import sampleRenderHook from '@hippotastic/sample-render-hook'
import netlify from '@astrojs/netlify/functions'

export default defineConfig({
  integrations: [sampleRenderHook({
    justTestingOptions: true,
    stringArray: ['a', 'b', 'c'],
  })],
  adapter: netlify(),
})
```

## How to develop a render hook integration

For development purposes, I've just put the source of my demo integration into `/packages/sample-render-hook`, so I need to add a link to `package.json` and also allow experimental integrations.

**package.json:**
```js
{
  // ...
  "scripts": {
    "dev": "astro dev --experimental-integrations",
    "start": "astro dev --experimental-integrations",
    "build": "astro build --experimental-integrations",
  },
  "devDependencies": {
    // ...
    "@hippotastic/sample-render-hook": "./packages/sample-render-hook",
    // ...
  },
  // ...
}
```

This is the hook implementation itself, based on your React integration:

**packages/sample-render-hook/package.json:**
```json
{
  "name": "@hippotastic/sample-render-hook",
  "description": "Use render hooks within Astro",
  "version": "0.1.0",
  "type": "module",
  "author": "hippotastic",
  "license": "MIT",
  "private": true,
  "exports": {
    ".": "./index.js",
    "./hook.js": "./hook.js"
  }
}
```

**packages/sample-render-hook/index.js:**
```js
function getRenderHook (options) {
  return {
    name: '@hippotastic/sample-render-hook',
    hookEntrypoint: '@hippotastic/sample-render-hook/hook.js',
    options,
  }
}

function src_default (options) {
  return {
    name: '@hippotastic/sample-render-hook',
    hooks: {
      'astro:config:setup': ({ addRenderHook }) => {
        addRenderHook(getRenderHook(options))
      },
    },
  }
}

export { src_default as default }
```

**packages/sample-render-hook/hook.js:**
```js
function postProcessHtml({ routeType, pathname, html, ssr, options }) {
  console.log(`*** Received postProcessHtml hook call! pathname="${pathname}", options=${JSON.stringify(options)}`)

  return `<!--START-->\n${html}\n<!--END-->`
}

export default {
  postProcessHtml,
}
```

This has been tested in dev, build and SSR (using Netlify). Please let me know how to proceed with this. Should I create a PR now? I've not done this yet because I've noticed that you've closed my previous PR, so I'm waiting for your input on the RFC process. 🙂

**Reply by FredKSchott** [reply]
Nice! This is super exciting! I'm surprised (in a good way!) that this worked so effortlessly. How do the `justTestingOptions:true` options work in SSR? I thought that the config file didn't exist in SSR (it's only used to build your project) so I would have thought that when you go to run it, `sample-render-hook/hook.js` is called without options.

**Reply by FredKSchott** [reply]
> Should I create a PR now? I've not done this yet because I've noticed that you've closed my previous PR, so I'm waiting for your input on the RFC process. 🙂

We have an RFC process that we follow for larger changes like this, that this will definitely fall under. It forces all of us to think through the implementation first, before writing the code PR (although a proof of concept like you have is always a good idea at this stage).

Basically, we're doing all the right things so far by talking it through here, identifying early issues, and confirming that there is interest from the rest of the team. I'd like to get other peoples feedback on this as well, especially @matthewp who knows how our SSR works better than anyone.

If this gets a 👍 from the team, then the next step is writing up an official RFC. You can learn more about our process here: https://github.com/withastro/rfcs

**Reply by hippotastic** [reply]
> How do the `justTestingOptions:true` options work in SSR? I thought that the config file didn't exist in SSR (it's only used to build your project) so I would have thought that when you go to run it, `sample-render-hook/hook.js` is called without options.

They get serialized along with the other properties of the hook and end up in the SSR bundle like this:
```js
const renderHooks = [Object.assign({"name":"@hippotastic/sample-render-hook","hookEntrypoint":"@hippotastic/sample-render-hook/hook.js","options":{"justTestingOptions":true,"stringArray":["a","b","c"]}}, { ssr: _renderHook0 }),];
```

I've replicated the logic that was already there for loading the renderers, and adapted it to make it work for the render hooks as well. The SSR bundle does not need access to the config file that way.

**Reply by FredKSchott** [reply]
smart!

**Comment by matthewp** [comment]
I really like this use case! I echo @FredKSchott's comments on some of the challenges, but I don't think they are insurmountable.

One thing that immediately comes to mind reading this is that this is a little like middleware. I could see these use cases which are somewhat related:

- Hook on incoming requests
  - Validate the structure
  - fetch a logged in user's profile, etc.
  - This is traditional middleware
- Hooks on outgoing responses
  - Security hooks, ie add `Strict-Transport-Security` header on all responses
  - HTML minifying, etc (this issue)

I wonder if we should first create lower-level `astro:request` and `astro:response` hooks and then layer HTML-specific hooks on later (which would provide the AST as explained above).

You would still have the ability to do HTML manipulation in a hypothetical `astro:response` hook(s), you would just be dealing with the raw bytes and not an AST.

This is also nice because we could pass this hypothetical hook something like a ReadableStream for the body and the hook would opt-in to buffering the HTML rather than Astro always doing so (and therefore disabling streaming).

The we could layer some HTML-specific hooks on later which provide the same abilities but with a nicer interface for the sort of use-cases discussed here.

**Reply by hippotastic** [reply]
Regarding the challenges mentioned by Fred, I think my implementation linked in https://github.com/withastro/rfcs/discussions/169#discussioncomment-2545371 already solved them and works without any issues in SSR. So luckily, this does not seem to be a problem.

I really like the idea of having a generic middleware API though! Being able to "plug in" middleware that e.g. takes care of user authentication with secure tokens would be a great developer-friendly way of extending the capabilities of Astro and reusing code, and something that developers are already used to from other frameworks which is currently missing in Astro. I can also definitely see that my use cases mentioned in this discussion could be solved by outgoing response middleware hooks.

However, I'd also like to point out that transforming the HTML output is a very typical use case which is easy to implement (as I think I've demonstrated) and I would love to get such functionality out to developers rather sooner than later, while discussing and finalizing a more generic solution.

**Reply by wassfila** [reply]
@matthewp I had similar ideas about Astro html transformations, and I see two types of possible pipelines
- Astro rehype pipeline, that could fit Astro as a classical rehype plugin, it is expected to share memory, and can pass an ast from step to step with low overhead, but the issue with ast is that it's in one block, so does not fit streaming
- Astro middlewares pipeline, here we can start by first allowing the existing middleware to share a context or to pass custom attributes between middlewares in the resp and req objects, this case can run even on a separate edge server and enhance some components with e.g. low cost user specific customizations.

**Comment by florian-lefebvre** [comment]
This is now possible using middlewares: https://docs.astro.build/en/guides/middleware/
